### PR TITLE
Add CMake to the project

### PR DESCRIPTION
**Pull request by palmerc**
I believe I've improved upon the previous CMake PR enough to warrant a complete do-over. :) You'll notice that in addition to building beautifully with make or ninja, it also generates an Xcode project that maintains the existing folder structure and most of the settings.

### CODE REVIEW

File: Clutch/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+set( SOURCE_OPTOOL
+        optool.h
+        optool-defines.h
+        optool-headers.h
+        optool-headers.m
+        optool-operations.h
+        optool-operations.m
+        NSData+Reading.h
+        NSData+Reading.m )
+source_group( "Supporting Files\\optools" FILES ${SOURCE_OPTOOL} )
```
Comment (Tatsh) [inline Clutch/CMakeLists.txt]: `/` you mean right?

**Comment by Tatsh** [comment]
Not working for me on latest Xcode:

```
~/dev/clutch/build
 $ cmake ..
-- The C compiler identification is AppleClang 9.1.0.9020039
-- The CXX compiler identification is AppleClang 9.1.0.9020039
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
MOBILECORESERVICES
    linked by target "Clutch" in directory /Users/tatsh/dev/clutch/Clutch
UIKIT
    linked by target "Clutch" in directory /Users/tatsh/dev/clutch/Clutch

-- Configuring incomplete, errors occurred!
See also "/Users/tatsh/dev/clutch/build/CMakeFiles/CMakeOutput.log".
```

I expect this to be able to work with only `cmake ..` (for a release build).

**Comment by palmerc** [comment]
It isn’t finding your Frameworks.

MOBILECORESERVICES and UIKit
Have you installed Xcode in an unusual place?

**Comment by palmerc** [comment]
I believe you invoked CMake incorrectly. You must specify a toolchain

cmake -G Xcode -D CMAKE_TOOLCHAIN_FILE=../Clutch/cmake/iphoneos.toolchain.cmake ../Clutch

**Comment by Tatsh** [comment]
Getting this error regarding minimum iOS version (set to 11.4):

```
clang: error: invalid version number in '-miphoneos-version-min='
```

Shouldn't [this](https://github.com/KJCracks/Clutch/pull/220/files#diff-e9d39ad9b4286f56b06cf693711fbe6bR158) be something else? Also it should be version 8.0 as the minimum.

Also [here](https://github.com/KJCracks/Clutch/pull/220/files#diff-612f98de3e1e69c434286956d64a3a8fR20).

**Comment by Tatsh** [comment]
With fixed paths, it looks like this works. However I still want the minimum iOS to be 8.0.

**Comment by Tatsh** [comment]
This is the command I am using now, from the Clutch root directory:

```bash
rm -fR build && mkdir build && cd build && cmake -G Xcode -D CMAKE_TOOLCHAIN_FILE=../cmake/iphoneos.toolchain.cmake ../Clutch && xcodebuild
```

With this, I get the `-miphoneos-version-min=` error.

**Comment by palmerc** [comment]
Little uncertain how you’re messing up the iOS Deployment version. Look in the CMakeLists.txt to find the following value

set( MACOSX_DEPLOYMENT_TARGET 8.0 )

When compiling with Ninja or Make

-miphoneos-version-min=${MACOSX_DEPLOYMENT_TARGET}
And when using Xcode

XCODE_ATTRIBUTE_IPHONEOS_DEPLOYMENT_TARGET ${MACOSX_DEPLOYMENT_TARGET}

**Comment by palmerc** [comment]
Andrew, are you guys on Slack or similar?

**Comment by palmerc** [comment]
Little uncertain how you’re not getting the correct iOS Deployment version. Look in the CMakeLists.txt to find the following value

    set( MACOSX_DEPLOYMENT_TARGET 8.0 )

When building with Ninja or Make you'll get the following flag added to the `target_compile_options`

    -miphoneos-version-min=${MACOSX_DEPLOYMENT_TARGET}

And when using Xcode the following is added to the project:

    XCODE_ATTRIBUTE_IPHONEOS_DEPLOYMENT_TARGET ${MACOSX_DEPLOYMENT_TARGET}

**Comment by palmerc** [comment]
OK, I've seen what you're seeing. When building at the command-line with `xcodebuild` it is passing `-miphoneos-version-min=11.4` and `-O0` which says to me Debug and that I need to poke around in the settings to determine why it is making these choices.

**Comment by palmerc** [comment]
I went ahead and created a POST_BUILD step that replaces the bash script `move_and_sign.sh`. So we'll copy the binary into the `build/` folder and call code sign on it.

**Comment by Tatsh** [comment]
Finally got a chance to test this. Merging now.

**Pull request merged**